File: app/src/main/java/android/academy/nytimes/network/RetrofitHelper.java
package android.academy.nytimes.network;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitHelper {

    private static final String BASE_URL = "http://api.nytimes.com/svc/topstories/v2/";
    private static volatile RetrofitHelper INSTANCE;
    static final String TOKEN = "aae03167237f4fbf836838d292db06a8";

    private RetrofitHelper() {
    }

    public static RetrofitHelper getInstance() {
        RetrofitHelper localInstance = INSTANCE;
        if (localInstance == null) {
            synchronized (RetrofitHelper.class) {
                localInstance = INSTANCE;
                if (localInstance == null) {
                    localInstance = INSTANCE = new RetrofitHelper();
                }
            }
        }
        return localInstance;
    }

    NYTimesApi getNYTimesApi() {
        Gson gson = new GsonBuilder()
                .setLenient()
                .create();

        return new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()
                .create(NYTimesApi.class);
    }
}

File: app/src/main/java/android/academy/nytimes/network/RetrofitHelper.kt
package android.academy.nytimes.network

import com.google.gson.GsonBuilder
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

class RetrofitHelper private constructor(private val url: String,
                                         private val interceptor: Interceptor) {

    companion object {
        @Volatile
        private var INSTANCE: RetrofitHelper? = null

        fun getInstance(url: String, interceptor: Interceptor): RetrofitHelper {
            var localInstance = INSTANCE
            if (localInstance == null) {
                synchronized(RetrofitHelper::class.java) {
                    localInstance = INSTANCE
                    if (localInstance == null) {
                        INSTANCE = RetrofitHelper(url, interceptor)
                        localInstance = INSTANCE
                    }
                }
            }
            return localInstance!!
        }
    }

    fun getNyTimesApi(): NYTimesApi = with(Retrofit.Builder()) {
        baseUrl(url)
        addConverterFactory(GsonConverterFactory.create(createGson()))
        addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        client(createOkHttpClient())
        build()
    }.create(NYTimesApi::class.java)


    private fun createGson() = with(GsonBuilder()) {
        setLenient()
        serializeNulls()
        create()
    }

    private fun createOkHttpClient() = with(OkHttpClient.Builder()) {
        addInterceptor(interceptor)
        build()
    }
}
